selfdrive/pandad/pandad.cc

```diff
@@ -487,27 +456,23 @@ void panda_state_thread(std::vector<Panda *> pandas, bool spoofing_started) {
     for (const auto &panda : pandas) {
       panda->send_heartbeat(engaged);
     }
-
-    rk.keepTime();
   }
 }
 
 
-void peripheral_control_thread(Panda *panda, bool no_fan_control) {
-  util::set_thread_name("pandad_peripheral_control");
+// run at 20 HZ
+void process_peripheral_state(Panda *panda, PubMaster *pm, bool no_fan_control) {
+  static SubMaster sm({"deviceState", "driverCameraState"});
 
-  SubMaster sm({"deviceState", "driverCameraState"});
+  static uint64_t last_driver_camera_t = 0;
+  static uint16_t prev_fan_speed = 999;
+  static uint16_t ir_pwr = 0;
+  static uint16_t prev_ir_pwr = 999;
 
-  uint64_t last_driver_camera_t = 0;
-  uint16_t prev_fan_speed = 999;
-  uint16_t ir_pwr = 0;
-  uint16_t prev_ir_pwr = 999;
-
-  FirstOrderFilter integ_lines_filter(0, 30.0, 0.05);
-
-  while (!do_exit && panda->connected()) {
-    sm.update(1000);
+  static FirstOrderFilter integ_lines_filter(0, 30.0, 0.05);
 
+  {
+    sm.update(0);
```

Comment:
`sm.update` should probably happen in the main pandad_run loop for readability

Reply:
There are two submasters updated at different frequencies. For now, keeping updates in their own functions provides better encapsulation. Moving to pandad_run would add complexity. We can address this in a future refactor.